pkg/controller/compute/agentpool/managed.go

```diff
@@ -0,0 +1,161 @@
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+	azureclients "github.com/crossplane/provider-azure/pkg/clients"
+	"github.com/crossplane/provider-azure/pkg/clients/compute/agentpool"
+)
+
+// Error strings.
+const (
+	errNotAgentPool    = "managed resource is not a AgentPool"
+	errCreateAgentPool = "cannot create AgentPool"
+	errGetAgentPool    = "cannot get AgentPool"
+	errDeleteAgentPool = "cannot delete AgentPool"
+)
+
+// SetupAgentPool adds a controller that reconciles AgentPools.
+func SetupAgentPool(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(v1alpha3.AgentPoolGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha3.AgentPool{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha3.AgentPoolGroupVersionKind),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, auth, err := azure.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	cl := containerservice.NewAgentPoolsClient(creds[azureclients.CredentialsKeySubscriptionID])
+	cl.Authorizer = auth
+	return &external{c: cl}, nil
+}
+
+type external struct {
+	c containerserviceapi.AgentPoolsClientAPI
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha3.AgentPool)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAgentPool)
+	}
+
+	c, err := e.c.Get(ctx, cr.Spec.AgentPoolParameters.ResourceGroupName, cr.Spec.AgentPoolParameters.AKSClusterName, meta.GetExternalName(cr))
+	if azure.IsNotFound(err) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetAgentPool)
+	}
+
+	agentpool.Update(cr, &c)
+	needUpdate := agentpool.NeedUpdate(cr, &c)
+
+	if cr.Status.State != "Succeeded" {
+		return managed.ExternalObservation{
+			ResourceExists: true,
+			// Restrict updates in unstable state
+			ResourceUpToDate: true,
+		}, nil
+	}
+
+	cr.SetConditions(xpv1.Available())
+
+	o := managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: !needUpdate,
+	}
+	return o, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha3.AgentPool)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotAgentPool)
+	}
+	cr.SetConditions(xpv1.Creating())
```

Comment:
crossplane-runtime can now [handle setting "creating" and "deleting" conditions](https://github.com/crossplane/crossplane-runtime/pull/242).